vshard/util.lua

```diff
@@ -286,6 +286,15 @@ local function fiber_is_self_canceled()
     return not pcall(fiber.testcancel)
 end
 
+--
+-- Check if connection's fiber was explicitly killed
+--
+local conn_fiber_killed = function(conn)
+    local msg = 'fiber is cancelled'
+    return (conn.state == 'error_reconnect' or conn.state == 'error') and
```

Comment:
Probably you can avoid string comparison here...

Could `conn.state == 'error_reconnect' or conn.state == 'error')` be replaced with `conn.error ~= nil`?
Also probably error sometimes not table but `box.error` that has `code`. And comparison of codes is faster than string comparison.

Reply:
Unfortunately, they don't have code. In 2.7 - 2.10 `conn.error` is indeed not a table but a cdata. However, it's not a `box.error` but an exception, described in lib/core/exception.cc:
```C
static const struct method_info exception_methods[] = {
	make_method(&type_Exception, "message", &Exception::get_errmsg),
	make_method(&type_Exception, "log", &Exception::log),
	METHODS_SENTINEL
};
const struct type_info type_Exception = make_type("Exception", NULL,
	exception_methods);

<...>

const struct type_info type_FiberIsCancelled =
	make_type("FiberIsCancelled", &type_Exception);

FiberIsCancelled::FiberIsCancelled(const char *file, unsigned line)
	: Exception(&type_FiberIsCancelled, file, line)
{
	error_format_msg(this, "fiber is cancelled");
}

```
It has nothing inside it but this message. That's probably the reason, why now this structure is not used as an error and is replaced by the message. So, we have to compare error's strings if we want this issue to be solved from vshard. 
However, it's not that bad as in case of replacing `conn.state == 'error_reconnect' or conn.state == 'error'` with just `conn.error` strings won't be compared in most cases.

Reply:
> Unfortunately, they don't have code. In 2.7 - 2.10 conn.error is indeed not a table but a cdata. However, it's not a box.error but an exception, described in lib/core/exception.cc:

exception itself can't be lua object (even cdata). There is wrapper around such errors - https://github.com/tarantool/tarantool/blob/master/src/lua/error.lua

Reply:
It will be anyway comparison of strings (either it will be `base_type`, `type`, or `message`):

<details>
<summary>Tarantool  2.6 </summary>

```lua
tarantool> _TARANTOOL
---
- 2.6.0-0-g47aa4e01e
...
tarantool> c
---
- peer_uuid: 98d4e7e7-2602-4ce3-857d-82fcda96eb40
  host: 127.0.0.1
  schema_version: 78
  protocol: Binary
  state: error
  error: fiber is cancelled
  peer_version_id: 132608
  port: '3301'
...
tarantool> c.error:unpack()
---
- base_type: FiberIsCancelled
  type: FiberIsCancelled
  message: fiber is cancelled
  trace:
  - file: ./src/lua/fiber.c
    line: 46
...
```
</details>

<details>
<summary>Tarantool  2.9 </summary>

```lua
tarantool> _TARANTOOL
---
- 2.9.0-0-g15bbdab92
...
tarantool> c
---
- peer_uuid: 59b9898b-be3a-440b-b385-6224472d1894
  host: 127.0.0.1
  schema_version: 78
  protocol: Binary
  state: error
  error: fiber is cancelled
  peer_version_id: 133376
  port: '3301'
...
tarantool> c.error:unpack()
---
- base_type: FiberIsCancelled
  type: FiberIsCancelled
  message: fiber is cancelled
  trace:
  - file: /build/tarantool/src/tarantool/src/lua/fiber.c
    line: 46
...
```
</details>